### DISCUSSION TITLE
Sperrung des Buttons Metadaten holen beim DOI-Import

### DISCUSSION THREAD

**Discussion by vgerlach** [OPUS4/community#21] [General]
Es sollte überlegt werden, ob man das Eingabefeld für die DOI nach dem Druck auf ‚Metadaten holen‘ sperrt – zumindest für den Fall, dass der Datenimport erfolgreich war, d.h. dass ein valider DOI eingegeben und prozessiert wurde. 
Hintergrund dazu ist - neben der Frage ob es funktional Sinn macht mehrere DOIs nacheinander eingeben zu können - , dass derzeit die Eingabefelder nicht zurückgesetzt werden, wenn man mehrere DOIs nacheinander verwendet. D.h. wenn die erste DOI z.B. 5 Autoren hat und die nächste DOI nur 2, dann werden 5 Autorenfelder angezeigt (nur 2 befüllt, was ok ist) - das Formular wird also nicht zurückgesetzt.

**Comment by j3nsch** [comment]
@alw-bsz Oder sollte das als Bug behandelt werden? Das Formular sollte zurückgesetzt werden, wenn die DOI geändert wird. Die Umsetzung wird sich in Zukunft, mit einem neuen Publish-Modul, noch mal ändern und das kann dann dabei berücksichtigt werden. Die Sperrung könnte eine kurzfristige Lösung sein.

**Comment by alw-bsz** [comment]
Sperren des DOI-Eingabefeldes halte ich für unnötig und "user-unfriedly". Man kann sich mal vertippen und versehentlich die falschen Daten abrufen. Dann sollte man das auch einfach korrigieren können, ohne nochmals den Veröffentlichungsprozess neu starten zu müssen. Ich würde klar eine Lösung bevorzugen, bei der beim Klick auf den Metadaten-abrufen-Button die Fomular-Felder zunächst zurückgesetzt und anschließend neu befüllt werden.

**Comment by bfalkenstein** [comment]
Das Thema der "überflüssigen" Felder hat mich schon während der Entwicklung des DOI-Imports beschäftigt. Ich habe jedoch keine Möglichkeit gefunden, mittels JavaScript das Formular tatsächlich komplett zu resetten (außer durch einen neuen Veröffentlichungsvorgang).  Deshalb wird es nun lediglich geleert, was zu den genannten überzähligen Feldern führt. Da diese jedoch beim Abspeichern ignoriert werden, fand ich diese Lösung praktikabel. 

Die Eingabe mehrerer DOIs hintereinander ist einerseits wegen möglicher Tippfehler erlaubt, andererseits auch, um nach einer erfolglosen Suche ("DOI wurde nicht bei Crossref gefunden") mit einer anderen DOI weitermachen zu können, ohne den Veröffentlichungsvorgang nochmal komplett neu starten zu müssen. 
Die Idee, das DOI-Feld für Änderungen nur dann zu sperren, wenn ein Import erfolgreich durchgeführt wurde, wäre technisch umsetzbar. Das entsprechende Flag (doi_import_successful = true) ist ja bereits vorhanden und könnte dafür genutzt werden. 
Damit wäre dann das o.g. Weiterarbeiten nach erfolgloser Suche weiterhin möglich. Allerdings würde eine falsch eingegebe DOI dann nicht mehr "verziehen", eine Korrektur wäre nicht möglich.

**Comment by j3nsch** [comment]
Eine Lösung mit der aktuellen Implementation wäre sicherlich möglich, aber wahrscheinlich unnötig kompliziert. Wenn ich es richtig verstehe werden im Augenblick beim Import dynamisch weitere INPUT-Elemente hinzugefügt und befüllt, ein POST zum Server geschickt und damit das Formular angepasst neu aufgebaut. Ich habe mir das nicht im Detail angeschaut. Umgekehrt könnte man vermutlich auch überflüssige Formular-Elemente entfernen, einen POST abschicken und damit das Formular wieder reduziert aufbauen. Evlt. stören dabei die Daten in der Session auf dem Server. Einfacher wäre vermutlich ein "Reset"-Request, der den initialen Aufruf der zweiten Formularseite simuliert ohne noch einmal das erste Formular zu bemühen.

Letztendlich ist aber klar, das das Publish-Modul neu implementiert werden muss. Damit ändert sich auch noch einmal die Implementation des DOI-Imports. Alles was wir jetzt machen wäre temporär. Mit einer neuen Implementation würde der Import vermutlich auf die Server-Seite wandern. Dort kann einfach für jeden neuen Import ein komplett neues Formular aufgebaut werden. Ein POST findet ja auch jetzt schon statt. Der Ablauf dann wäre also POST nur mit DOI zum Server, Server Code holt und mapped Metadaten (alles PHP), Formular wird angezeigt.
